### PR TITLE
adds 1st party certificate extension support

### PR DESCRIPTION
**Pull request by andrewpmartinez**
Reworked the controller a little and simplified this greatly from what it was.

### CODE REVIEW

File: inc_internal/authenticators.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "zt_internal.h"
+#include "utils.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern void ziti_send_event(ziti_context ztx, const ziti_event_t *e);
```
Comment (ekoby) [inline inc_internal/authenticators.h:28]: this belongs in zt_internal.h

File: library/CMakeLists.txt [hunk 2]
```diff
@@ -35,7 +35,7 @@ SET(ZITI_INCLUDE_DIRS
     PRIVATE ${uv-mbed_SOURCE_DIR}/src
 )
 
-add_library(ziti     STATIC ${ZITI_SRC_FILES} ${ZITI_HEADER_FILES} )
+add_library(ziti     STATIC ${ZITI_SRC_FILES} ${ZITI_HEADER_FILES} authenticators.c)
```
Comment (ekoby) [inline library/CMakeLists.txt:38]: put this in `ZITI_SRC_FILES`

File: library/ziti_ctrl.c [hunk 3]
```diff
@@ -835,4 +835,57 @@ void ziti_ctrl_post_mfa_recovery_codes(ziti_controller *ctrl, char *body, size_t
...
+    resp->body_parse_func = (int (*)(void *, const char *, size_t)) parse_ziti_extend_cert_authenticator_resp_ptr;
+    resp->resp_cb = (void (*)(void *, const ziti_error *, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = ctrl_default_cb;
+
+    char path[128];
+    snprintf(path, sizeof(path), "/current-identity/authenticators/%s/extend", authenticatorId);
+
+    NEWP(extend_req, ziti_extend_cert_authenticator_req);
```
Comment (ekoby) [inline library/ziti_ctrl.c:853]: don't need to allocate, make it a stack var

File: library/ziti_ctrl.c [hunk 4]
```diff
@@ -835,4 +835,57 @@ void ziti_ctrl_post_mfa_recovery_codes(ziti_controller *ctrl, char *body, size_t
...
+    resp->body_parse_func = NULL;
+    resp->resp_cb = (void (*)(void *, const ziti_error *, void *)) cb;
+    resp->ctx = ctx;
+    resp->ctrl = ctrl;
+    resp->ctrl_cb = ctrl_default_cb;
+
+    char path[256];
+    snprintf(path, sizeof(path), "/current-identity/authenticators/%s/extend-verify", authenticatorId);
+
+    NEWP(verify_req, ziti_verify_extend_cert_authenticator_req);
```
Comment (ekoby) [inline library/ziti_ctrl.c:879]: same as above

File: library/authenticators.c [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    .error = err,
+                    .is_verified = false,
+                    .authenticator_id = authenticator_ctx->authenticator_id,
+            }
+    };
+
+    ziti_send_event(ztx, &ev);
+}
+
+const char* CAN_NOT_UPDATE_AUTHENTICATOR = "CAN_NOT_UPDATE_AUTHENTICATOR";
```
Comment (RodneyRichardson) [inline library/authenticators.c:42]: Should CAN_NOT be one word: CANNOT?

File: inc_internal/authenticators.h (hunk 1 above)
Comment (andrewpmartinez) [reply inc_internal/authenticators.h:28]: Fixed.

File: library/CMakeLists.txt (hunk 2 above)
Comment (andrewpmartinez) [reply library/CMakeLists.txt:38]: Fixed

File: library/authenticators.c (hunk 5 above)
Comment (andrewpmartinez) [reply library/authenticators.c:42]: Quick googling shows that when stating something is not possible both forms are correct. 

I am inclined to take no action.

File: library/ziti_ctrl.c (hunk 3 above)
Comment (andrewpmartinez) [reply library/ziti_ctrl.c]: fixed

File: library/ziti_ctrl.c (hunk 4 above)
Comment (andrewpmartinez) [reply library/ziti_ctrl.c]: fixed

File: library/ziti_ctrl.c [hunk 6]
```diff
@@ -844,4 +844,52 @@ void ziti_ctrl_post_mfa_recovery_codes(ziti_controller *ctrl, char *body, size_t
...
+
+    size_t body_len;
+    char *body = ziti_extend_cert_authenticator_req_to_json(&extend_req, 0, &body_len);
+
+    um_http_req_t *req = start_request(ctrl->client, "POST", path, ctrl_resp_cb, resp);
+    um_http_req_header(req, "Content-Type", "application/json");
+    um_http_req_data(req, body, body_len, free_body_cb);
+}
+
+void ziti_ctrl_verify_extend_cert_authenticator(ziti_controller *ctrl, char *authenticatorId, char *client_cert, void(*cb)(void *, const ziti_error *, void *), void *ctx) {
```
Comment (ekoby) [inline library/ziti_ctrl.c:873]: `authenticatorId`, `client_cert` should be `const`

File: library/ziti_ctrl.c [hunk 7]
```diff
@@ -844,4 +844,52 @@ void ziti_ctrl_post_mfa_recovery_codes(ziti_controller *ctrl, char *body, size_t
     um_http_req_t *req = start_request(ctrl->client, "POST", "/current-identity/mfa/recovery-codes", ctrl_resp_cb, resp);
     um_http_req_header(req, "Content-Type", "application/json");
     um_http_req_data(req, body, body_len, free_body_cb);
+}
+
+void ziti_ctrl_extend_cert_authenticator(ziti_controller *ctrl, char *authenticatorId, char *csr, void(*cb)(ziti_extend_cert_authenticator_resp*, const ziti_error *, void *), void *ctx) {
+    if(!verify_api_session(ctrl, cb, ctx)) return;
```
Comment (ekoby) [inline library/ziti_ctrl.c:850]: if there is an error, callback should be called. Caller may need to free resources

File: library/ziti_ctrl.c [hunk 8]
```diff
@@ -844,4 +844,52 @@ void ziti_ctrl_post_mfa_recovery_codes(ziti_controller *ctrl, char *body, size_t
...
+    size_t body_len;
+    char *body = ziti_extend_cert_authenticator_req_to_json(&extend_req, 0, &body_len);
+
+    um_http_req_t *req = start_request(ctrl->client, "POST", path, ctrl_resp_cb, resp);
+    um_http_req_header(req, "Content-Type", "application/json");
+    um_http_req_data(req, body, body_len, free_body_cb);
+}
+
+void ziti_ctrl_verify_extend_cert_authenticator(ziti_controller *ctrl, char *authenticatorId, char *client_cert, void(*cb)(void *, const ziti_error *, void *), void *ctx) {
+    if(!verify_api_session(ctrl, cb, ctx)) return;
```
Comment (ekoby) [inline library/ziti_ctrl.c:874]: same as above

File: library/ziti_ctrl.c (hunk 6 above)
Comment (andrewpmartinez) [reply library/ziti_ctrl.c:873]: Done

File: includes/ziti/error_defs.h [hunk 9]
```diff
@@ -77,21 +77,27 @@ limitations under the License.
...
-    /** returned when authentication is attempted but there is an existing api session waiting for auth queries to pass */                    \
-    XX(PARTIALLY_AUTHENTICATED, "api session is partially authenticated, waiting for auth query resolution") \
+    /** not found, usually indicates stale reference or permission */ \
+    XX(NOT_FOUND, "entity no longer exists or is no longer accessible") \
+    /** operation attempted while ziti_context is not enabled */ \
+    XX(DISABLED, "ziti context is disabled") \
+    /** returned when authentication is attempted but there is an existing api session waiting for auth queries to pass */ \
+    XX(PARTIALLY_AUTHENTICATED, "api session is partially authenticated, waiting for auth query resolution")               \
+    /** returned during certificate authenticator extension processing if an extendable authenticator is not found */\
+    XX(COULD_NOT_RETREIVE_AUTHENTICATOR, "a valid certificate based authenticator could not be retrieved")                 \
```
[the remark this answers is not in the document]
Comment (andrewpmartinez) [reply includes/ziti/error_defs.h:96]: Someone else got to it before me. Thanks!

File: includes/ziti/ziti.h [hunk 10]
```diff
@@ -1002,6 +1002,54 @@ ZITI_FUNC
...
+ * Responses are provided via `ziti_extend_certificate_authenticator_event` events. On that event
+ * check the `error` field for issues. If there are no errors persist the `new_client_cert_pem` field
+ * and make a subsequent call to `ziti_verify_extend_cert_authenticator` to enable the new client certificate.
+ *
+ * @param ztx the handle to the Ziti Edge identity context needed for other Ziti C SDK functions
+ * @param csr a CSR representing the request for a new client certificate
+ * @param ctx additional context to be passed back in raised events
+ */
+ZITI_FUNC
+extern void ziti_extend_cert_authenticator(ziti_context ztx, char *csr_pem, void *ctx);
```
Comment (ekoby) [inline includes/ziti/ziti.h:1020]: `const csr_pem`

File: includes/ziti/ziti.h [hunk 11]
```diff
@@ -1002,6 +1002,54 @@ ZITI_FUNC
...
+ * `ziti_extend_certificate_authenticator_event` with `is_verified` as true or errors set in the `error` field.
+ *
+ * `ziti_set_client_cert` should be called shortly after on success.
+ *
+ * @param ztx the handle to the Ziti Edge identity context needed for other Ziti C SDK functions
+ * @param new_cert the new client certificate that will become active on successful verification, provided in the extension event
+ * @param ctx additional context to be passed back in raised events
+ */
+ZITI_FUNC
+extern void ziti_verify_extend_cert_authenticator(ziti_context ztx, char *new_cert, void *ctx);
```
Comment (ekoby) [inline includes/ziti/ziti.h:1039]: const new_cert

Comment (andrewpmartinez) [reply includes/ziti/ziti.h:1039]: Done

File: includes/ziti/ziti.h (hunk 10 above)
Comment (andrewpmartinez) [reply includes/ziti/ziti.h:1020]: Done

File: library/authenticators.c [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+        ziti_send_event(wrapped_ctx->ztx, &ev);
+    }
+
+    FREE(wrapped_ctx->authenticator_id);
+    FREE(wrapped_ctx);
+}
+
+void ziti_extend_cert_authenticator(ziti_context ztx, const char* csr_pem, void *ctx) {
+    ZTX_LOG(INFO, "attempting to extend certificate authenticator id: %s", ztx->api_session->authenticator_id);
```
Comment (ekoby) [inline library/authenticators.c:122]: all new functions should check that api_session is present and return  code (ZITI_INVALID_STATE) if not
this code will crash if ztx is in unauthenticated state

you'll need to change the signature to return `int` as it is the only way to communicate error

Comment (andrewpmartinez) [reply library/authenticators.c:85]: Just to confirm, for example you want void -> int on the return type:

````
int ziti_extend_cert_authenticator(...);
```

And then something like:

```
int ziti_extend_cert_authenticator(...) {
    if(ztx->api_session == NULL){
       return ZITI_INVALID_STATE;
    }
    
    ...
    
    return ZITI_OK;
}
```

File: includes/ziti/ziti.h [hunk 13]
```diff
@@ -1001,6 +1001,78 @@ extern void ziti_mfa_auth(ziti_context ztx, const char *code, ziti_mfa_cb auth_c
...
+ * @brief Extend Certificate Authenticator Callback
+ *
+ * This callback is invoked on the conclusion of ziti_extend_cert_authenticator(). The result of the function
+ * may be an error condition so it is important to verify the status code in this callback. In the
+ * event the certificate authenticator extension request cannot be completed an error will be returned otherwise
+ * #ZITI_OK is expected.
+ *
+ * @see ziti_extend_cert_authenticator(), ZITI_ERRORS
+ */
+typedef void (*ziti_extend_cert_authenticator_cb)(ziti_context ztx, const char* authenticator_id, const char* cert_pem, int status, void *ctx);
```
Comment (ekoby) [inline includes/ziti/ziti.h:1014]: what purpose `authenticator_id` serves here -- there is no function that takes it?
 is the app supposed to use it?

File: library/authenticators.c (hunk 12 above)
Comment (ekoby) [reply library/authenticators.c:85]: we need to either return an error code or call `cb` with it (in which case the function can still be `void`)
my preference is to return the error to be consistent with other functions

File: library/ziti.c [hunk 14]
```diff
@@ -170,6 +170,10 @@ int load_tls(ziti_config *cfg, tls_context **ctx) {
     return ZITI_OK;
 }
 
+void ziti_set_client_cert(ziti_context ztx, const char *cert_buf, size_t cert_len, const char *key_buf, size_t key_len) {
+    ztx->tlsCtx->api->set_own_cert(ztx->tlsCtx, cert_buf, cert_len, key_buf, key_len);
```
Comment (ekoby) [inline library/ziti.c:174]: TLS API `set_own_cert` may return an error if key/cert in invalid format or don't match

File: includes/ziti/ziti.h [hunk 15]
```diff
@@ -1001,6 +1001,78 @@ extern void ziti_mfa_auth(ziti_context ztx, const char *code, ziti_mfa_cb auth_c
...
+ * @param ctx additional context to be passed back in raised events
+ */
+ZITI_FUNC
+extern void ziti_verify_extend_cert_authenticator(ziti_context ztx, const char *new_cert, ziti_verify_extend_cert_authenticator_cb cb, void *ctx);
+
+/**
+ * @brief Updates the certificate context for the ZTX with a new client certificate and key.
+ *
+ * @param ztx the handle to the Ziti Edge identity context needed for other Ziti C SDK functions
+ * @param cert_buf a x509 certificate
```
Comment (ekoby) [inline includes/ziti/ziti.h:1069]: I believe cert and  key have to be in PEM format

File: includes/ziti/ziti.h (hunk 13 above)
Comment (andrewpmartinez) [reply includes/ziti/ziti.h:1014]: Nothing other than logging info, but it doesn't need to be in the CB if the c-sdk logs it. Will remove.

File: library/ziti.c (hunk 14 above)
Comment (andrewpmartinez) [reply library/ziti.c:174]: Sure thing, will route an error out and log some context here.

File: library/authenticators.c (hunk 12 above)
Comment (andrewpmartinez) [reply library/authenticators.c:85]: Addressed.

File: library/ziti.c (hunk 14 above)
Comment (andrewpmartinez) [reply library/ziti.c:174]: I just return the result of the internal call. I am not sure what all the non-zero error return meanings are.

File: includes/ziti/ziti.h (hunk 15 above)
Comment (andrewpmartinez) [reply includes/ziti/ziti.h:1069]: Clarified.

**Pull request merged**